### PR TITLE
Add debounce for playground code updates in order to reduce lag

### PR DESCRIPTION
**Pull request by samuelclements**
Added a debounce to the playground code updates in order to reduce the lag effect when updating larger components with a lot of text typing at once.

Have also wrapped the playground code in a React Fragment in order to allow for adjacent jsx tags and empty strings.

### CODE REVIEW

File: website/src/utils/ast.js [hunk 1]
```diff
@@ -4,6 +4,10 @@ import generate from "@babel/generator";
 import * as t from "@babel/types";
 import { allComponentNames } from "./meta";
 
+function wrapCodeInFragment(code) {
+  return `<React.Fragment>${code}</React.Fragment>`;
```
[the remark this answers is not in the document]
Comment (samuelclements) [reply website/src/utils/ast.js]: Yeah good point, the problem is that it is the babel `parse()` function that is returning an error when it isn't wrapped, so I never actually get the AST to traverse :( 

I've tested with a function and it doesn't seem to break it when wrapped in a `React.Fragment` but probably isn't ideal. I'll try and see if I can find another way to do it.

**Comment by samuelclements** [comment]
@moroshko I've updated the playground code wrapping to only wrap if the code does not start with a class or a function. Let me know what you think.

File: website/src/utils/ast.test.js [hunk 2]
```diff
@@ -104,3 +104,44 @@ describe("getReactLiveNoInline", () => {
...
+        render() {
+          return (
+            <Button>Button</Button>
+          )
+        }
+      }`
+    );
+
+    expect(functionShorthand.startsWith("<React.Fragment>")).toBe(false);
+    expect(functionShorthand.endsWith("</React.Fragment>;")).toBe(false);
```
Comment (moroshko) [inline website/src/utils/ast.test.js]: This feels simpler to me:
```
expect(annotateCodeForPlayground(myFunc)).toBe(myFunc);
```

File: website/src/utils/ast.js [hunk 3]
```diff
@@ -4,6 +4,16 @@ import generate from "@babel/generator";
 import * as t from "@babel/types";
 import { allComponentNames } from "./meta";
 
+function codeContainsClassOrFunction(code) {
+  const trimmedCode = typeof code === "string" && code.trim();
```
[the remark this answers is not in the document]
Comment (samuelclements) [reply website/src/utils/ast.js]: Yeah true, it actually should always be a string as code is defaulted to `""` when we set state. I'll remove this check 👍

File: website/src/utils/ast.test.js [hunk 4]
```diff
@@ -104,3 +104,72 @@ describe("getReactLiveNoInline", () => {
...
+    expect(annotateCodeForPlayground(functionShorthand).startsWith("<>")).toBe(
+      false
+    );
+    expect(annotateCodeForPlayground(functionShorthand).endsWith("</>;")).toBe(
+      false
+    );
+  });
+
+  it("does not wrap code starting with 'function' Fragment", () => {
+    const functionLonghand = "function App() { <Button>Apply</Button> }";
```
Comment (moroshko) [inline website/src/utils/ast.test.js]: I assume you wanted to return that JSX?

Comment (samuelclements) [reply website/src/utils/ast.test.js]: Yes I did ;) Updated now 👍

File: website/src/utils/ast.test.js [hunk 5]
```diff
@@ -104,3 +104,72 @@ describe("getReactLiveNoInline", () => {
...
 });
+
+describe("annotateCodeForPlayground", () => {
+  it("wraps plain jsx code in Fragment", () => {
+    expect(
+      annotateCodeForPlayground("<Button>Apply</Button>").startsWith("<>")
+    ).toBe(true);
+    expect(
+      annotateCodeForPlayground("<Button>Apply</Button>").endsWith("</>;")
+    ).toBe(true);
```
Comment (moroshko) [inline website/src/utils/ast.test.js]: IMHO, this feels too verbose. If we test plain JSX, we should state it only once. Having 2 copies is prone to errors. Also, seeing `startsWith` and `endsWith` in these tests make them less readable (harder to scan).
Usually, I wouldn't add helper methods in tests, but I think here the cost of not doing so is too high.
```
function isFragment(code) {
  return /^<>.*<\/>;$/.test(code);
}

it("wraps plain jsx code in Fragment", () => {
  expect(isFragment(annotateCodeForPlayground("<Button>Apply</Button>"))).toBe(true);
});
```

What do you think?

Comment (samuelclements) [reply website/src/utils/ast.test.js]: Yeah I think it's worth it to add in a helper method seeing as I'm using the `startsWith` and `endsWith` a few times. I generally find regex's pretty hard to read when scanning as well but this one is probably simple enough.

**Comment by moroshko** [comment]
@samuelclements  Thanks for fixing this!

**Pull request merged**